File: II semester/HomeWork4/Zadacha1 ParseTree/Zadacha1 ParseTree/Tree.cs
using System;

namespace ParseTreeNamespace
{
    /// <summary>
    /// Class tree.
    /// </summary>
    public class Tree
    {
        /// <summary>
        /// Head of tree.
        /// </summary>
        public TreeElement head { get; set; }

        /// <summary>
        /// Is element number?
        /// </summary>
        /// <param name="element"></param>
        /// <returns></returns>
        private bool IsNumber(char element)
        {
            return (((int)element >= (int)('0')) && ((int)element - (int)('0') <= 9));
        }

        /// <summary>
        /// Is element function?
        /// </summary>
        /// <param name="element"></param>
        /// <returns></returns>
        private bool IsFunction(char element)
        {
            return ((element == '+') | (element == '-') | (element == '*') | (element == '/'));
        }

        private Operation NewOperation(char value)
        {
            Operation theOperation;
            if (value == '+')
            {
                theOperation = new OperationSum();
            }
            else if (value == '-')
            {
                theOperation = new OperationDiff();
            }
            else if (value == '*')
            {
                theOperation = new OperationMulti();
            }
            else
            {
                theOperation = new OperationDiv();
            }
            theOperation.left = null;
            theOperation.right = null;
            theOperation.up = null;
            return theOperation;
        }

        /// <summary>
        /// Insert sub tree by expression and index-variable.
        /// </summary>
        /// <param name="expression"></param>
        /// <param name="index"></param>
        /// <param name="element"></param>
        private void InsertSubTree(string expression, ref int index, TreeElement element)
        {
            while (index < expression.Length)
            {
                char symbol = expression[index];
                if (symbol == '(')
                {
                    TreeElement newElement = NewOperation(expression[index + 1]);
                    newElement.up = element;
                    if (element == null)
                    {
                        this.head = newElement;
                        element = newElement;
                    }
                    else
                    {
                        if (element.left == null)
                        {
                            element.left = newElement;
                        }
                        else
                        {
                            element.right = newElement;
                        }
                        element = newElement;
                    }
                    index += 2;
                    InsertSubTree(expression, ref index, element);
                    InsertSubTree(expression, ref index, element);
                }
                else
                {
                    if (IsNumber(expression[index]))
                    {
                        TreeElement newTreeElement = new Operand(element, (int)expression[index] - (int)'0');
                        index++;
                    }
                    else
                    {
                        if (expression[index] == ')')
                        {
                            element = element.up;
                            index++;
                        }
                    }
                }
            }
        }

        /// <summary>
        /// Write in tree expression.
        /// </summary>
        /// <param name="expression"></param>
        public void WriteInTree(string expression)
        {
            TreeElement element = this.head;
            int i = 0;
            InsertSubTree(expression, ref i, element);
        }

        /// <summary>
        /// Print tree.
        /// </summary>
        /// <returns></returns>
        public string Print()
        {
            string result = "";
            this.head.Print(ref result);
            return result;
        }

        /// <summary>
        /// Return value of the expression-tree.
        /// </summary>
        /// <returns></returns>
        public int ValueOfExpression()
        {
            TreeElement element = this.head;
            if (element != null)
            {
                return (element.Count());
            }
            return 0;
        }
    }
}

File: I semester/HomeWork7/Zadacha3 Parse Tree/Zadacha3/tree.cpp
#include <iostream>
#include "tree.h"

using namespace std;

struct TreeElement
{
	Element value;
	Number number;
	TreeElement *left;
	TreeElement *right;
	TreeElement *up;
};

struct Tree
{
	TreeElement *head;
};

Tree *create()
{
	Tree *newTree = new Tree;
	newTree->head = nullptr;
	return newTree;
}

void deleteSubTree(TreeElement *element)
{
	if (element != nullptr)
	{
		deleteSubTree(element->left);
		deleteSubTree(element->right);
		delete element;
	}
}

void deleteTree(Tree *tree)
{
	deleteSubTree(tree->head);
	delete tree;
}

TreeElement *createNewElement(Element value)
{
	TreeElement *temp = new TreeElement;
	temp->value = value;
	temp->number = 0;
	temp->left = nullptr;
	temp->right = nullptr;
	temp->up = nullptr;
	return temp;
}

Position first(Tree *tree)
{
	return tree->head;
}

Position left(Tree *tree, Position position)
{
	return position->left;
}

Position right(Tree *tree, Position position)
{
	return position->right;
}

Position up(Tree *tree, Position position)
{
	return position->up;
}

bool isNumber(char element)
{
	return ((static_cast<int>(element) >= static_cast<int>('0')) && (static_cast<int>(element) - static_cast<int>('0') <= 9));
}

bool isFunction(char element)
{
	return ((element == '+') | (element == '-') | (element == '*') | (element == '/'));
}

void createNewElementWithNumber(Tree *tree, Position position, char value)
{
	TreeElement *newElement = new TreeElement;
	newElement->number = static_cast<int>(value) - static_cast<int>('0');
	newElement->left = nullptr;
	newElement->right = nullptr;
	newElement->value = '#';
	newElement->up = position;
	if (position->left == nullptr)
	{
		position->left = newElement;
	}
	else
	{
		position->right = newElement;
	}
}

void insertSubTree(Tree *tree, char *buffer, int &i, Position position)
{
	while (i < strlen(buffer))
	{
		char p = buffer[i];
		if (p == '(')
		{
			TreeElement *newElement = createNewElement(buffer[i + 1]);
			newElement->up = position;
			if (position == nullptr)
			{
				tree->head = newElement;
				position = newElement;
			}
			else
			{
				if (position->left == nullptr)
				{
					position->left = newElement;
				}
				else
				{
					position->right = newElement;
				}
				position = newElement;
			}
			i += 2;
			insertSubTree(tree, buffer, i, position);
			insertSubTree(tree, buffer, i, position);
		}
		else
		{
			if (isNumber(buffer[i]))
			{
				createNewElementWithNumber(tree, position, buffer[i]);
				i++;
			}
			else
			{
				if (buffer[i] == ')')
				{
					position = up(tree, position);
					i++;
				}
			}
		}
	}
}

void writeInTree(Tree *tree, char *buffer)
{
	Position position = first(tree);
	int i = 0;
	insertSubTree(tree, buffer, i, position);
}

void printSubTree(Position position)
{
	if (isFunction(position->value))
	{
		cout << "( " << position->value << " ";
		printSubTree(position->left);
		printSubTree(position->right);
	}
	else
	{
		cout << position->number << " ";
		if (position == position->up->right)
			cout << ") ";
	}
}

void print(Tree *tree)
{
	Position position = tree->head;
	if (position != nullptr)
		printSubTree(position);
	else
		cout << "Tree is clear!\n";
}

int makeOperation(int number1, int number2, char operation)
{
	if (operation == '+')
	{
		return (number1 + number2);
	}
	if (operation == '*')
	{
		return (number1 * number2);
	}
	if (operation == '-')
	{
		return (number1 - number2);
	}
	if (operation == '/')
	{
		return (number1 / number2);
	}
}

int count(Position position)
{
	if (isFunction(position->value))
	{
		return (makeOperation(count(position->left), count(position->right), position->value));
	}
	else
	{
		return (position->number);
	}
}

int valueOfExpression(Tree *tree)
{
	Position position = tree->head;
	if (position != nullptr)
	{
		return (count(position));
	}
	return 0;
}